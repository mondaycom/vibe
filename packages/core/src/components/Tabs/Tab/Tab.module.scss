@import "~monday-ui-style/dist/mixins";
@import "../../../styles/states";
@import "../../../styles/typography";

.tabWrapper {
  position: relative;
  display: inline-block;
  padding-left: 1px;
  padding-right: 1px;
  border-top: 1px solid transparent;
  border-bottom: 2px solid;
  border-bottom-color: var(--ui-background-color);
  color: var(--primary-text-color);
  text-align: center;
  height: 100%;
  margin: 0;
}

.tabWrapper:after {
  display: block;
  position: absolute;
  left: 0;
  right: 0;
  content: "";
  border-bottom: solid 2px;
  border-bottom-color: var(--primary-color);
  transform: scaleX(0);
  z-index: 0;
}

.tabWrapper .tabInner {
  $margin-bottom: 1px;
  height: calc(100% - #{$margin-bottom});
  display: flex;
  align-items: center;
  justify-content: center;
  text-decoration: none;
  padding: 4px 16px;
  @include vibe-text(text1, normal);
  user-select: none;
  margin-bottom: $margin-bottom;
  cursor: pointer;
  color: var(--primary-text-color);
}

.tabWrapper .tabInner:focus {
  outline: none;
}

.tabWrapper .tabInner .tabIcon {
  color: var(--icon-color);
  margin-right: var(--spacing-small);
}

.tabWrapper .tabInner .tabIcon.right {
  margin-right: 0;
  margin-left: var(--spacing-small);
}

.tabWrapper.tabFocusVisibleInset {
  @include focus-style-css-inset(3px, 3px);
  position: relative;
}

.tabWrapper.active:after {
  transform: scaleX(1);
  transition: transform var(--motion-productive-medium) var(--motion-timing-enter);
}

.tabWrapper.disabled .tabInner {
  color: var(--disabled-text-color);
  cursor: not-allowed;
}

.tabWrapper:not(.disabled) .tabInner:hover {
  border-radius: 4px;
  background-color: var(--primary-background-hover-color);
}
