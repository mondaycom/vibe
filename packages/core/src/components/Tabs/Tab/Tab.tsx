import cx from "classnames";
import React, { FC, forwardRef, ReactElement, useRef } from "react";
import { noop as NOOP } from "lodash-es";
import useMergeRef from "../../../hooks/useMergeRef";
import { getStyle } from "../../../helpers/typesciptCssModulesHelper";
import Icon from "../../Icon/Icon";
import VibeComponentProps from "../../../types/VibeComponentProps";
import { IconType } from "../../Icon";
import { ComponentDefaultTestId, getTestId } from "../../../tests/test-ids-utils";
import styles from "./Tab.module.scss";
import { SubIcon } from "../../../types/SubIcon";
import Tooltip, { TooltipProps } from "../../Tooltip/Tooltip";
import { ComponentVibeId } from "../../../tests/constants";
import { keyCodes } from "../../../constants";

export interface TabProps extends VibeComponentProps {
  /**
   * Class name applied to the inner tab content.
   */
  tabInnerClassName?: string;
  /**
   * The index value of the tab.
   */
  value?: number;
  /**
   * If true, disables the tab.
   */
  disabled?: boolean;
  /**
   * If true, marks the tab as active.
   */
  active?: boolean;
  /**
   * If true, applies focus styles to the tab.
   */
  focus?: boolean;
  /**
   * The icon displayed in the tab.
   */
  icon?: SubIcon;
  /**
   * The type of icon.
   */
  iconType?: IconType;
  /**
   * The position of the icon relative to the text.
   */
  iconSide?: string;
  /**
   * Callback fired when the tab is clicked.
   */
  onClick?: (value: number) => void;
  /**
   * Props passed to the tab's tooltip.
   */
  tooltipProps?: Partial<TooltipProps>;
  /**
   * The content displayed inside the tab.
   */
  children?: string | ReactElement | ReactElement[];
}

const Tab: FC<TabProps> = forwardRef(
  (
    {
      className,
      tabInnerClassName,
      id,
      value = 0,
      disabled = false,
      active = false,
      focus = false,
      onClick = NOOP,
      tooltipProps = {} as TooltipProps,
      icon,
      iconType,
      iconSide = "left",
      children,
      "data-testid": dataTestId
    }: TabProps,
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRef(ref, componentRef);

    function renderIconAndChildren() {
      if (!icon) return children;

      const iconElement = (
        <Icon
          ariaHidden={true}
          iconType={iconType}
          icon={icon}
          className={cx(styles.tabIcon, getStyle(styles, iconSide))}
          iconSize={18}
          ignoreFocusStyle
        />
      );

      const childrenArray = React.Children.toArray(children);

      if (iconSide === "left") {
        return [iconElement, ...childrenArray];
      }

      return [...childrenArray, iconElement];
    }

    function handleKeyDown(event: React.KeyboardEvent) {
      if (event.key === keyCodes.ENTER || event.key === keyCodes.SPACE) {
        event.preventDefault();
        !disabled && onClick(value);
      }
    }

    return (
      <Tooltip {...tooltipProps} content={tooltipProps.content}>
        <li
          ref={mergedRef}
          key={id}
          className={cx(styles.tabWrapper, className, {
            [styles.active]: active,
            [styles.disabled]: disabled,
            [styles.tabFocusVisibleInset]: focus
          })}
          id={id}
          role="tab"
          aria-selected={active}
          aria-disabled={disabled}
          data-testid={dataTestId || getTestId(ComponentDefaultTestId.TAB, id)}
          data-vibe={ComponentVibeId.TAB}
          onClick={() => !disabled && onClick(value)}
          onKeyDown={handleKeyDown}
        >
          <div className={cx(styles.tabInner, tabInnerClassName)}>{renderIconAndChildren()}</div>
        </li>
      </Tooltip>
    );
  }
);

export default Tab;
