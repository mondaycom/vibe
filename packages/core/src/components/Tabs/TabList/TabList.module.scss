@import "../../../styles/typography";

.tabsWrapper .tabsList {
  display: flex;
  flex-direction: row;
  list-style-type: none;
  font-size: 0;
  position: relative;
  height: 40px;
  outline: none;
  padding: 0;
}

.tabsWrapper .tabsList.sm {
  height: 32px;
}

.tabsWrapper .tabsList.lg {
  height: 48px;
}

.tabsWrapper.stretched .tabsList {
  width: 100%;
}

.tabsWrapper.stretched .tabsList .tabListTabWrapper {
  width: 100%;
}

.tabsWrapper.stretched .tabsList .tabListTabWrapper .tabListTabInner {
  width: 100%;
}

.tabsWrapper.disabled {
  opacity: 0;
  height: 0;
  padding-bottom: 5px;
  pointer-events: none;
}

.tabsWrapper.stretchedUnderline {
  border-bottom: 2px solid var(--ui-background-color);
}
