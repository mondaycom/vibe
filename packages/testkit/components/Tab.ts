import { Page, Locator } from "@playwright/test";
import { BaseElement } from "./BaseElement";

/**
 * Class representing a Tab element.
 * Extends the BaseElement class.
 */
export class Tab extends BaseElement {
  /**
   * Create a Tab.
   * @param {Page} page - The Playwright page object.
   * @param {Locator} locator - The locator for the Tab element.
   * @param {string} elementReportName - The name for reporting purposes.
   */
  constructor(page: Page, locator: Locator, elementReportName: string) {
    super(page, locator, elementReportName);
  }
}
