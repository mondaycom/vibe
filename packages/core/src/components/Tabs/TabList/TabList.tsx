import cx from "classnames";
import { camelCase } from "lodash-es";
import React, { FC, forwardRef, ReactElement, useCallback, useEffect, useMemo, useRef, useState } from "react";
import useGridKeyboardNavigation from "../../../hooks/useGridKeyboardNavigation/useGridKeyboardNavigation";
import useMergeRef from "../../../hooks/useMergeRef";
import usePrevious from "../../../hooks/usePrevious";
import VibeComponentProps from "../../../types/VibeComponentProps";
import { NOOP } from "../../../utils/function-utils";
import { TabProps } from "../Tab/Tab";
import { ComponentDefaultTestId, getTestId } from "../../../tests/test-ids-utils";
import { getStyle } from "../../../helpers/typesciptCssModulesHelper";
import styles from "./TabList.module.scss";

export interface TabListProps extends VibeComponentProps {
  onTabChange?: (tabId: number) => void;
  activeTabId?: number;
  tabType?: string;
  size?: string;
  children?: ReactElement<TabProps>[];
}

const TabList: FC<TabListProps> = forwardRef(
  (
    {
      className,
      id,
      onTabChange = NOOP,
      activeTabId = 0,
      tabType = "Compact",
      size,
      children,
      "data-testid": dataTestId
    },
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRef(ref, componentRef);

    const [activeTabState, setActiveTabState] = useState<number>(activeTabId);

    const prevActiveTabIdProp = usePrevious(activeTabId);

    useEffect(() => {
      // Update active tab if changed from props
      if (activeTabId !== prevActiveTabIdProp && activeTabId !== activeTabState) {
        setActiveTabState(activeTabId);
      }
    }, [activeTabId, prevActiveTabIdProp, activeTabState, setActiveTabState]);

    const disabledTabIds = useMemo(() => {
      const disabledIds = new Set<number>();
      React.Children.forEach(children, (child, index) => {
        if (child.props.disabled) {
          disabledIds.add(index);
        }
      });
      return disabledIds;
    }, [children]);

    const onTabSelect = useCallback(
      (tabId: number) => {
        if (disabledTabIds.has(tabId)) return;
        setActiveTabState(tabId);
        onTabChange && onTabChange(tabId);
      },
      [onTabChange, disabledTabIds]
    );

    const onTabClick = useCallback(
      (value: HTMLElement | void, tabId: number) => {
        const tabCallbackFunc = children[tabId].props?.onClick;
        if (disabledTabIds.has(tabId)) return;
        if (tabCallbackFunc) tabCallbackFunc(tabId);
        onTabSelect(tabId);
      },
      [children, disabledTabIds, onTabSelect]
    );

    const getItemByIndex = useCallback((index: number): ReactElement<TabProps> => children[index], [children]);
    const disabledIndexes = useMemo(() => Array.from(disabledTabIds), [disabledTabIds]);
    const ulRef = useRef();
    const { activeIndex: focusIndex, onSelectionAction } = useGridKeyboardNavigation({
      ref: ulRef,
      numberOfItemsInLine: children?.length,
      itemsCount: children?.length,
      getItemByIndex,
      onItemClicked: onTabClick,
      disabledIndexes
    });

    const tabsToRender = useMemo(() => {
      const childrenToRender = React.Children.map(children, (child, index) => {
        return React.cloneElement(child, {
          value: index,
          active: activeTabState === index,
          focus: focusIndex === index,
          onClick: onSelectionAction,
          className: cx(styles.tabListTabWrapper, child.props.className),
          tabInnerClassName: cx(styles.tabListTabInner, child.props.tabInnerClassName)
        });
      });
      return childrenToRender;
    }, [children, activeTabState, focusIndex, onSelectionAction]);

    return (
      <div
        ref={mergedRef}
        className={cx(styles.tabsWrapper, className, [getStyle(styles, camelCase(tabType))])}
        id={id}
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.TAB_LIST, id)}
      >
        <ul ref={ulRef} tabIndex={0} className={cx(styles.tabsList, [getStyle(styles, size)])} role="tablist">
          {tabsToRender}
        </ul>
      </div>
    );
  }
);

Object.assign(TabList, {
  isTabList: true
});

export default TabList;
