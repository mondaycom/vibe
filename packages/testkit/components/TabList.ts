import { Page, Locator, test } from "@playwright/test";
import { BaseElement } from "./BaseElement";
import { Tab } from "./Tab";

/**
 * Class representing a TabList element.
 * Extends the BaseElement class.
 */
export class TabList extends BaseElement {
  /**
   * Create a TabList.
   * @param {Page} page - The Playwright page object.
   * @param {Locator} locator - The locator for the TabList element.
   * @param {string} elementReportName - The name for reporting purposes.
   */
  constructor(page: Page, locator: Locator, elementReportName: string) {
    super(page, locator, elementReportName);
  }

  async getAllTabs(): Promise<Tab[]> {
    let tabs: Tab[] = [];
    await test.step(`Get all tabs in ${this.elementReportName}`, async () => {
      const tabsLocators = await this.locator.locator("li").all();
      const tabPromises = tabsLocators.map(
        async tabLocator => new Tab(this.page, tabLocator, await tabLocator.innerText())
      );
      tabs = await Promise.all(tabPromises);
    });
    return tabs;
  }

  /**
   * Get a tab by its name.
   * @param {string} tabName - The name of the button to retrieve.
   * @returns {Button} The button with the specified name.
   */
  async getTabByName(tabName: string): Promise<Tab | undefined> {
    let tab: Tab | undefined;
    await test.step(`Get tab by name ${tabName} in ${this.elementReportName}`, async () => {
      const tabLocator = this.locator.locator("li").filter({ hasText: tabName });
      tab = new Tab(this.page, tabLocator, `Tab: ${tabName}`);
    });
    return tab;
  }

  /**
   * Select a tab by its name.
   * @param {string} tabName - The name of the tab to select.
   * @returns {Promise<void>}
   */
  async selectTab(tabName: string): Promise<void> {
    await test.step(`Select tab ${tabName} in ${this.elementReportName}`, async () => {
      const tab = await this.getTabByName(tabName);
      if (tab) {
        await tab.click();
      } else {
        throw new Error(`Tab with name "${tabName}" not found in ${this.elementReportName}`);
      }
    });
  }
}
