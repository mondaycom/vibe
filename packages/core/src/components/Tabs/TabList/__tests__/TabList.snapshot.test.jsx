import { describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import TabList from "../TabList";
import Tab from "../../Tab/Tab";

describe("TabList renders correctly", () => {
  it("with children", () => {
    const tree = renderer
      .create(
        <TabList>
          <Tab>First</Tab>
        </TabList>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with className", () => {
    const tree = renderer
      .create(
        <TabList className="test">
          <Tab>First</Tab>
        </TabList>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with id", () => {
    const tree = renderer
      .create(
        <TabList id="test">
          <Tab>First</Tab>
        </TabList>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with activeTabId", () => {
    const tree = renderer
      .create(
        <TabList activeTabId={1}>
          <Tab>First</Tab>
          <Tab>Second</Tab>
        </TabList>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with size", () => {
    const tree = renderer
      .create(
        <TabList size="lg">
          <Tab>First</Tab>
          <Tab>Second</Tab>
        </TabList>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with tabType", () => {
    const tree = renderer
      .create(
        <TabList tabType="stretched">
          <Tab>First</Tab>
          <Tab>Second</Tab>
        </TabList>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with stretchedUnderline", () => {
    const tree = renderer
      .create(
        <TabList stretchedUnderline>
          <Tab>First</Tab>
          <Tab>Second</Tab>
        </TabList>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
