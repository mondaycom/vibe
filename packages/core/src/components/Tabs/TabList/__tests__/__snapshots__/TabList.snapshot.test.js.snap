// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabList renders correctly with activeTabId 1`] = `
<div
  className="tabsWrapper compact"
  data-testid="tab-list"
>
  <ul
    className="tabsList"
    role="tablist"
    tabIndex={0}
  >
    <li
      aria-disabled={false}
      aria-selected={false}
      className="tabWrapper tabListTabWrapper"
      data-testid="tab"
      data-vibe="Tab"
      role="tab"
    >
      <a
        className="tabInner tabListTabInner"
        onClick={[Function]}
      >
        First
      </a>
    </li>
    <li
      aria-disabled={false}
      aria-selected={true}
      className="tabWrapper tabListTabWrapper active"
      data-testid="tab"
      data-vibe="Tab"
      role="tab"
    >
      <a
        className="tabInner tabListTabInner"
        onClick={[Function]}
      >
        Second
      </a>
    </li>
  </ul>
</div>
`;

exports[`TabList renders correctly with children 1`] = `
<div
  className="tabsWrapper compact"
  data-testid="tab-list"
>
  <ul
    className="tabsList"
    role="tablist"
    tabIndex={0}
  >
    <li
      aria-disabled={false}
      aria-selected={true}
      className="tabWrapper tabListTabWrapper active"
      data-testid="tab"
      data-vibe="Tab"
      role="tab"
    >
      <a
        className="tabInner tabListTabInner"
        onClick={[Function]}
      >
        First
      </a>
    </li>
  </ul>
</div>
`;

exports[`TabList renders correctly with className 1`] = `
<div
  className="tabsWrapper test compact"
  data-testid="tab-list"
>
  <ul
    className="tabsList"
    role="tablist"
    tabIndex={0}
  >
    <li
      aria-disabled={false}
      aria-selected={true}
      className="tabWrapper tabListTabWrapper active"
      data-testid="tab"
      data-vibe="Tab"
      role="tab"
    >
      <a
        className="tabInner tabListTabInner"
        onClick={[Function]}
      >
        First
      </a>
    </li>
  </ul>
</div>
`;

exports[`TabList renders correctly with id 1`] = `
<div
  className="tabsWrapper compact"
  data-testid="tab-list_test"
  id="test"
>
  <ul
    className="tabsList"
    role="tablist"
    tabIndex={0}
  >
    <li
      aria-disabled={false}
      aria-selected={true}
      className="tabWrapper tabListTabWrapper active"
      data-testid="tab"
      data-vibe="Tab"
      role="tab"
    >
      <a
        className="tabInner tabListTabInner"
        onClick={[Function]}
      >
        First
      </a>
    </li>
  </ul>
</div>
`;

exports[`TabList renders correctly with size 1`] = `
<div
  className="tabsWrapper compact"
  data-testid="tab-list"
>
  <ul
    className="tabsList lg"
    role="tablist"
    tabIndex={0}
  >
    <li
      aria-disabled={false}
      aria-selected={true}
      className="tabWrapper tabListTabWrapper active"
      data-testid="tab"
      data-vibe="Tab"
      role="tab"
    >
      <a
        className="tabInner tabListTabInner"
        onClick={[Function]}
      >
        First
      </a>
    </li>
    <li
      aria-disabled={false}
      aria-selected={false}
      className="tabWrapper tabListTabWrapper"
      data-testid="tab"
      data-vibe="Tab"
      role="tab"
    >
      <a
        className="tabInner tabListTabInner"
        onClick={[Function]}
      >
        Second
      </a>
    </li>
  </ul>
</div>
`;

exports[`TabList renders correctly with tabType 1`] = `
<div
  className="tabsWrapper stretched"
  data-testid="tab-list"
>
  <ul
    className="tabsList"
    role="tablist"
    tabIndex={0}
  >
    <li
      aria-disabled={false}
      aria-selected={true}
      className="tabWrapper tabListTabWrapper active"
      data-testid="tab"
      data-vibe="Tab"
      role="tab"
    >
      <a
        className="tabInner tabListTabInner"
        onClick={[Function]}
      >
        First
      </a>
    </li>
    <li
      aria-disabled={false}
      aria-selected={false}
      className="tabWrapper tabListTabWrapper"
      data-testid="tab"
      data-vibe="Tab"
      role="tab"
    >
      <a
        className="tabInner tabListTabInner"
        onClick={[Function]}
      >
        Second
      </a>
    </li>
  </ul>
</div>
`;
